juju/juju
# Commit Branch Charm Config Update

## Description of change

This patch adds the crucial step of updating the charm configuration documents in the _settings_ collection when a branch is committed.

The settings documents are all written in a single transaction so that we never get into a partially applied state. 

This required adding a method to `State` in the settings module. I am not terribly fond of how this is done, but the generations requirements are exposing weaknesses in how settings is implemented in the state package.

## QA steps

- `export JUJU_DEV_FEATURE_FLAGS=generations`.
- Bootstrap.
- `juju deploy redis`.
- `juju branch test-branch`.
- `juju config redis password=pass --branch test-branch`.
- (Optional) `config redis password --branch master` and verify there is no value.
- `juju commit test-branch`.
- `juju config redis password` shows "pass".

## Documentation changes

None.

## Bug reference

N/A
